mantidproject/mantid
# Update gcc version to 12

See #37519 for full details, but conda-forge use `gcc` version 12, so this updates our version to match. The developer environment and the conda build config have both been changed.

A few compiler warnings appeared, which I've also fixed:
- In `Chunker.cpp` there was an include missing.
- In `Quasi.cpp` there was an unnecessary (accidental) copy due to the specified type of the loop variable.
- In `InstrumentWidgetMaskTab.cpp` the local variable `parent` shadowed a method also called `parent()`, and it generated a warning about being unintialised.
- In `SpectrumDetectorMappingTest.h` a couple of variables were not initialised.
- In `SumOverlappingTubes.cpp` and `MuonAlgorithmHelper.cpp` I've reused the existing vector to prevent an `array-bounds` warning.
- Test folders in `Framework/` with warnings have the `-Wno-maybe-uninitialized` compiler flag to stop warnings coming from the use of `EXPECT_CALL` with the `_` arguments, which is a google test construction.
- Disabled a couple of warnings for `stdsoap2.cpp`.
- Qt test executables now have the `-Wno-maybe-uninitialized` compiler flag to disable the `maybe-uninitialized` warning.

See [here](https://gcc.gnu.org/onlinedocs/gcc/gcc-command-options/options-to-request-or-suppress-warnings.html) for information on the various `gcc` warnings.

Fixes #37519 



### To test (Linux-only):
- Remove your `mantid-developer` environment with `conda env remove -n mantid-developer`
- Delete `build/CMakeCache.txt`
- Create your `mantid-developer` environment with `mamba env create -f mantid-developer-linux.yml`
- Check that your environment has the right version by running `conda list | grep gxx`
- Run `cmake --preset=linux`, the compiler version detected should be 12.3 (or 12.* anyway)
- Compile
- Go and get a cup of tea
- Everything should compile and run as normal

Check that the Linux CI builds used version 12 instead of 10 by looking at the build logs, they'll need a clean build to pick up the compiler change ~(I'm waiting until after 6.10 is released before triggering those)~





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
